hardware_interface/include/hardware_interface/resource_manager.hpp

```diff
@@ -0,0 +1,115 @@
+#include "hardware_interface/sensor_hardware.hpp"
+#include "hardware_interface/system_hardware_interface.hpp"
+#include "hardware_interface/system_hardware.hpp"
+#include "hardware_interface/types/hardware_interface_return_values.hpp"
+#include "hardware_interface/visibility_control.h"
+#include "pluginlib/class_loader.hpp"
+
+using hardware_interface::return_type;
+
+namespace resource_manager
+{
+
+class ResourceManager
+{
+public:
+  HARDWARE_INTERFACE_PUBLIC
+  ResourceManager();
+
+  HARDWARE_INTERFACE_PUBLIC
+  virtual
+  ~ResourceManager() = default;
+
+  //  Non real-time safe functions
+  HARDWARE_INTERFACE_PUBLIC
+  return_type load_and_configure_resources_from_urdf(std::string urdf_string);
+
+  HARDWARE_INTERFACE_PUBLIC
+  return_type check_command_interfaces(
+    const std::string & joint_name, const std::vector<std::string> & interfaces) const;
+
+  HARDWARE_INTERFACE_PUBLIC
+  return_type claim_command_handle(
+    const std::string & joint_name, const std::vector<std::string> & interfaces,
+    std::shared_ptr<hardware_interface::components::Joint> & command_handle);
+
+  //  Real-time safe functions
+  HARDWARE_INTERFACE_PUBLIC
+  return_type start_all_resources();
+
+  HARDWARE_INTERFACE_PUBLIC
+  return_type stop_all_resources();
+
+  HARDWARE_INTERFACE_PUBLIC
+  return_type read_all_resources();
+
+  HARDWARE_INTERFACE_PUBLIC
+  return_type write_all_resources();
+
+private:
+  // TODO(all): make this unique?
+  std::vector<std::shared_ptr<hardware_interface::ActuatorHardware>> actuators_;
+  std::vector<std::shared_ptr<hardware_interface::SensorHardware>> sensors_;
+  std::vector<std::shared_ptr<hardware_interface::SystemHardware>> systems_;
+
+  std::map<std::string, std::shared_ptr<
+      hardware_interface::components::Joint>> joint_components_;
+  std::map<std::string, std::vector<std::shared_ptr<
+      hardware_interface::components::Joint>>> joint_components_for_hardware_;
+  std::map<std::string, std::shared_ptr<
+      hardware_interface::components::Sensor>> sensor_components_;
+  std::map<std::string, std::vector<std::shared_ptr<
+      hardware_interface::components::Sensor>>> sensor_components_for_hardware_;
+
+  /**
+   * Map of joints and their command interfaces
+   */
+  std::map<std::string, std::vector<std::string>> command_interfaces_;
+
+  std::map<std::string, std::vector<std::string>> claimed_command_interfaces_;
+
+  /**
+   * Map of joints and their state interfaces
+   */
+  std::map<std::string, std::vector<std::string>> state_interfaces_;
+
+  std::shared_ptr<pluginlib::ClassLoader<
+      hardware_interface::ActuatorHardwareInterface>> actuator_loader_;
+  std::shared_ptr<pluginlib::ClassLoader<
+      hardware_interface::SensorHardwareInterface>> sensor_loader_;
+  std::shared_ptr<pluginlib::ClassLoader<
+      hardware_interface::SystemHardwareInterface>> system_loader_;
+
+  std::shared_ptr<pluginlib::ClassLoader<hardware_interface::components::Joint>> joint_loader_;
```

Comment:
see #170, #171 I feel like they components should have pure interfaces as well and behave similar to the hardware resources.

Reply:
I propose to change this later when #170 and #171 are getting merged.